common/infallible/src/nonzero.rs

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+/// A wrapper around `std::num::NonZeroUsize` to no longer worry about `unwrap()`
+#[macro_export]
+macro_rules! NonZeroUsize {
+    ($num:expr) => {
+        NonZeroUsize!($num, "Must be non-zero")
+    };
+    ($num:expr, $message:literal) => {
+        std::num::NonZeroUsize::new($num).expect($message)
```

Comment:
one big reason for `nonzero_ext` is to get compile time proof that a literal is nonzero. you can see the assertion here: https://docs.rs/nonzero_ext/0.2.0/src/nonzero_ext/lib.rs.html#311

that way you can write `nonzero!(123)` without any unwraps because it checks that the compile-time constant is nonzero. if you write `nonzero!(0)` it will error at compile time.

here we don't communicate that and instead just silently panic inside at runtime, which seems suboptimal.

Reply:
Yeah. I think that this change is more a step backward in terms of type safety.

Reply:
As I said earlier in Slack, if you look at where these are being used, its either A: a test, which is not that big of a difference, except the tests have to run to fail.  Or B: With information loaded from a Config, which is not possible to error at compile time, because it at startup of runtime.

I get your concern that this isn't quite as quick at telling us if a number is NonZero.  But it's added a lot of boilerplate in here.  Let's talk outside this.

Reply:
Good arguments.

* Unix has /dev/null but we don't as we don't allow zero sized queues
* The proposal as it stands would pull in another dependency for just a macro :/

I do love type safety so it might be worth bringing back something in the future if folks really want to enforce non-zero queues.